prow/config/branch_protection.go

```diff
@@ -176,15 +178,88 @@ type BranchProtection struct {
 	// ProtectTested determines if branch protection rules are set for all repos
 	// that Prow has registered jobs for, regardless of if those repos are in the
 	// branch protection config.
-	ProtectTested bool `json:"protect-tested-repos,omitempty"`
+	ProtectTested *bool `json:"protect-tested-repos,omitempty"`
 	// Orgs holds branch protection options for orgs by name
 	Orgs map[string]Org `json:"orgs,omitempty"`
 	// AllowDisabledPolicies allows a child to disable all protection even if the
 	// branch has inherited protection options from a parent.
-	AllowDisabledPolicies bool `json:"allow_disabled_policies,omitempty"`
+	AllowDisabledPolicies *bool `json:"allow_disabled_policies,omitempty"`
 	// AllowDisabledJobPolicies allows a branch to choose to opt out of branch protection
 	// even if Prow has registered required jobs for that branch.
-	AllowDisabledJobPolicies bool `json:"allow_disabled_job_policies,omitempty"`
+	AllowDisabledJobPolicies *bool `json:"allow_disabled_job_policies,omitempty"`
+}
+
+func isPolicyEmpty(p Policy) bool {
+	return apiequality.Semantic.DeepEqual(p, Policy{})
+}
+
+func (bp *BranchProtection) merge(additional *BranchProtection) error {
+	var errs []error
+	if !isPolicyEmpty(bp.Policy) && !isPolicyEmpty(additional.Policy) {
+		errs = append(errs, errors.New("both brachprotection configs set a top-level policy"))
+	} else {
+		bp.Policy = additional.Policy
```

Comment:
Is it expected to replace `bp.Policy` with empty `additional.Policy`?

Reply:
Do you mean if its expected that if both `bp.Policy` and `additional.Policy` are empty we will overwrite the former with the latter? If yes, well its not explicitly wanted but it effectively a no-op and not extra handling this case keeps the code simpler

Reply:
I meant when `bp.Policy` is not empty and `additional.Policy` is empty, the `else` condition is executed, which would override `bp.Policy` with the empty `additional.Policy`

Reply:
I might be wrong, `&&` condition for two negations is a little bit mind twisting :)

Reply:
Oh you are right, that was absolutely not intended, very good catch. Fixed it and added a `TestBranchprotectionMergeRemainsUnchangedWhenMergingEmptyIn`. 

Also renamed `isPolicyEmpty` to `isPolicySet` to avoid all the negations.

Only good thing about this messy code is that it's very easy to test via fuzzing/property based testing.